### PR TITLE
100% coverage

### LINKED ISSUE #14: Get unit test coverage to 100%
And pin it.

### PR DESCRIPTION
**Pull request by jamadden**
Fixes #14

Commits are usually for one file.

I discovered and fixed a few small bugs. 

I discovered a bug in formatting times with timezones (https://github.com/zopefoundation/zope.i18n/commit/376b5438893ec6b3f456ceb166407bffeacc1b7d) that I wasn't sure what to do about so I just left a test for the bug.

### CODE REVIEW

File: src/zope/i18n/__init__.py [hunk 1]
```diff
@@ -99,6 +131,29 @@ def translate(msgid, domain=None, mapping=None, context=None,
...
+
+    If no target language is provided, but a context is and we were able to
+    find a translation domain, we will use the `negotiate` function to
+    attempt to determine the language to translate to:
+
+    .. doctest::
+        :hide:
+
+        >>> from zope import i18n
+        >>> old_negatiate = i18n
```
Comment (mgedmin) [inline src/zope/i18n/__init__.py]: I think this was supposed to be `old_negotiate = i18n.negotiate`

File: src/zope/i18n/__init__.py [hunk 2]
```diff
@@ -99,6 +131,29 @@ def translate(msgid, domain=None, mapping=None, context=None,
...
+
+    >>> def test_negatiate(context):
+    ...     print("Negatiating for %r" % (context,))
```
Comment (mgedmin) [inline src/zope/i18n/__init__.py]: Typo: negatiate -> negotiate (in several places, with different capitalizations).

File: src/zope/i18n/compile.py [hunk 3]
```diff
@@ -1,54 +1,61 @@
-from contextlib import closing
+
 import logging
-import os
-from os.path import join
-from stat import ST_MTIME
+logger = logging.getLogger('zope.i18n')
+
+def _cannot_compile_mo_file(domain, lc_messages_path):
+    logger.critical("Unable to compile messages: Python `gettext` library missing.")
```
Comment (mgedmin) [inline src/zope/i18n/compile.py]: This message makes me think it's talking about https://docs.python.org/3/library/gettext.html, when it's actually talking about https://pypi.python.org/pypi/python-gettext.

File: src/zope/i18n/compile.py [hunk 4]
```diff
@@ -1,54 +1,61 @@
...
+        pofile = str(base + '.po')
+        mofile = str(base + '.mo')
 
-    po_mtime = 0
-    try:
-        po_mtime = os.stat(pofile)[ST_MTIME]
-    except (IOError, OSError):
-        return
+        po_mtime = _safe_mtime(pofile)
+        mo_mtime = _safe_mtime(mofile) if os.path.exists(mofile) else 0
```
Comment (mgedmin) [inline src/zope/i18n/compile.py]: It seems strange to me to distinguish the cases where the file doesn't exist and where it existed briefly but was gone in the next nanosecond.

File: src/zope/i18n/compile.py [hunk 5]
```diff
@@ -1,54 +1,61 @@
...
-            logger.warning('Error while compiling %s (%s).', pofile, err)
+        if po_mtime > mo_mtime:
+            try:
+                # Msgfmt.getAsFile returns io.BytesIO on Python 3, and cStringIO.StringIO
+                # on Python 2; sadly StringIO isn't a proper context manager, so we have to
+                # wrap it with `closing`. Also, Msgfmt doesn't properly close a file
+                # it opens for reading if you pass the path, but it does if you pass
+                # the file.
+                with open(pofile, 'rb') as pofd:
+                    with closing(Msgfmt(pofd, domain).getAsFile()) as mo:
```
Comment (mgedmin) [inline src/zope/i18n/compile.py]: Would be nice if the thing returned by Msgfmt.getAsFile() implemented `__enter__` and `__exit__` natively.

File: src/zope/i18n/compile.py [hunk 6]
```diff
@@ -1,54 +1,61 @@
...
     from pythongettext.msgfmt import Msgfmt
     from pythongettext.msgfmt import PoSyntaxError
-except ImportError:
+except ImportError: # pragma: no cover
     HAS_PYTHON_GETTEXT = False
+    compile_mo_file = _cannot_compile_mo_file
+else:
+    from contextlib import closing
+    import os.path
+    from os.path import join
```
Comment (mgedmin) [inline src/zope/i18n/compile.py]: python-gettext is a 20 KB Python library with no external dependencies and with Python 3 support.  I think we can depend on it and avoid the conditional checks.

(I'm not a fan of how all the code in this module was indented in this PR: this makes the diff hard to read.  I had to switch to the Split view in GitHub and then switch between `?w=1`, which shows the changes but doesn't let me comment, and no `?w=1`, which lets me comment, but where I have to compare the functions by eyeball.)

File: src/zope/i18n/compile.py [hunk 7]
```diff
@@ -1,54 +1,61 @@
...
+                        with open(mofile, 'wb') as fd:
+                            fd.write(mo.read())
+                # For testing we return distinct values for each scenario
+                return True
+            except PoSyntaxError as err:
+                logger.warning('Syntax error while compiling %s (%s).', pofile, err.msg)
+                return 0
+            except (IOError, OSError) as err:
+                logger.warning('Error while compiling %s (%s).', pofile, err)
+                return False
```
Comment (mgedmin) [inline src/zope/i18n/compile.py]: I don't like returning `0` vs `False`.  It's meant for tests, but I'm sure actual user code will start doing `is 0` checks to distinguish syntax errors from other kinds of errors.

File: src/zope/i18n/config.py [hunk 8]
```diff
@@ -1,11 +1,37 @@
...
+    >>> _parse_languages(None) is None
+    True
+    >>> _parse_languages("en") == frozenset(('en',))
+    True
+    >>> _parse_languages("en,es") == frozenset(('en', 'es'))
+    True
+    """
+    if value:
+        value = value.strip().replace(",", " ")
+        value = frozenset(value.split())
```
Comment (mgedmin) [inline src/zope/i18n/config.py]: The `.strip()` is not necessary: `str.split()` ignores leading and trailing whitespace.

File: src/zope/i18n/config.py [hunk 9]
```diff
@@ -1,11 +1,37 @@
...
+    """
+    Utility function to parse languages.
+
+    >>> _parse_languages(None) is None
+    True
+    >>> _parse_languages("en") == frozenset(('en',))
+    True
+    >>> _parse_languages("en,es") == frozenset(('en', 'es'))
+    True
+    """
```
Comment (mgedmin) [inline src/zope/i18n/config.py]: I'd like to see the doctests indented four spaces.

I'd also like to see at least one test case that has whitespace in it.

File: src/zope/i18n/format.py [hunk 10]
```diff
@@ -467,8 +471,8 @@ def format(self, obj, pattern=None, rounding=True):
         if bin_pattern[PADDING2] is not None and pre_padding > 0:
             if bin_pattern[PADDING1] is not None:
                 text += bin_pattern[PADDING2]
-            else:
-                text += bin_pattern[PADDING2]*pre_padding
+            else: # pragma: no cover
+                text += bin_pattern[PADDING2] * pre_padding
```
Comment (mgedmin) [inline src/zope/i18n/format.py]: Too hard to test?

File: src/zope/i18n/format.py [hunk 11]
```diff
@@ -659,17 +666,18 @@ def buildDateTimeInfo(dt, calendar, pattern):
     # Getting the timezone right
     tzinfo = dt.tzinfo or pytz.utc
     tz_secs = tzinfo.utcoffset(dt).seconds
-    tz_secs = (tz_secs > 12*3600) and tz_secs-24*3600 or tz_secs
+    tz_secs = tz_secs - 24 * 3600 if tz_secs > 12 * 3600 else tz_secs
```
Comment (mgedmin) [inline src/zope/i18n/format.py]: (Unrelated to this PR) I have doubts about the correctness of this logic.  Time zones [UTC+12:45, UTC+13:00, UTC+14:00 exist](https://en.wikipedia.org/wiki/List_of_UTC_time_offsets#UTC+12:45,_M*).

File: src/zope/i18n/locales/inheritance.py [hunk 12]
```diff
@@ -178,6 +178,12 @@ class InheritingDictionary(Inheritance, dict):
       [1, 2, 3]
       >>> list(locale.data.items())
       [(1, 'eins'), (2, 'two'), (3, 'three')]
+
+    We currently fail to override ``values``, but instead inherited
+    data can be seen in the ``value`` method::
+
+      >>> sorted(locale.data.value())
+      ['eins', 'three', 'two']
```
Comment (mgedmin) [inline src/zope/i18n/locales/inheritance.py]: Is that a bug?

File: src/zope/i18n/locales/inheritance.py [hunk 13]
```diff
@@ -206,18 +212,20 @@ def get(self, name, default=None):
...
     def keys(self):
-        return [item[0] for item in self.items()]
+        return list(self._make_reified_inherited_dict().keys())
 
     def value(self):
-        return [item[1] for item in self.items()]
-
-
+        # XXX: Was this meant to override values() and is just a typo?
+        return list(self._make_reified_inherited_dict().values())
```
Comment (mgedmin) [inline src/zope/i18n/locales/inheritance.py]: I bet it was.

File: src/zope/i18n/locales/xmlfactory.py [hunk 14]
```diff
@@ -1251,6 +1275,15 @@ def _extractDelimiters(self):
...
+          ...   <identity>
+          ...     <version number="1.0"/>
+          ...     <generation date="2003-12-19" />
+          ...     <language type="en" />
+          ...     <territory type="US" />
+          ...     <variant type="POSIX" />
+          ...   </identity>
+          ... </ldml>'''
+
           ... </ldml>'''
```
Comment (mgedmin) [inline src/zope/i18n/locales/xmlfactory.py]: This looks wrong.

File: src/zope/i18n/tests/test_compile.py [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+from zope.i18n import compile
+
+
+@unittest.skipUnless(compile.HAS_PYTHON_GETTEXT,
+                     "Need python-gettext")
+class TestCompile(unittest.TestCase):
+
+    def test_non_existant_path(self):
+
+        self.assertIsNone(compile.compile_mo_file('no_such_domain', ''))
```
Comment (mgedmin) [inline src/zope/i18n/tests/test_compile.py]: The unnecessary blank line in this tiny method bugs me.

File: src/zope/i18n/tests/test_compile.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+    def test_po_exists_but_invalid(self):
+        import tempfile
+        import shutil
+        import os.path
+
+        td = tempfile.mkdtemp()
```
Comment (mgedmin) [inline src/zope/i18n/tests/test_compile.py]: Please pass `prefix='test-zope.i18n-` to `mkdtemp()`.

EDIT: or a `suffix='.zope18n_test_...` if you prefer.

File: src/zope/i18n/tests/test_compile.py [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+        self.addCleanup(shutil.rmtree, td)
+
+        with open(os.path.join(td, "foo.po"), 'w') as f:
+            f.write("this should not compile")
+
+        self.assertEqual(
+            0,
+            compile.compile_mo_file('foo', td))
```
Comment (mgedmin) [inline src/zope/i18n/tests/test_compile.py]: Besides, if you were using 0 vs False to distinguish outcomes in the unit tests, well, `0 == False` so ...

File: src/zope/i18n/tests/test_compile.py [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+        td = tempfile.mkdtemp(suffix=".zopei18n_test_compile")
+        self.addCleanup(shutil.rmtree, td)
+
+        mofile = os.path.join(td, 'foo.mo')
+        with open(mofile, 'w') as f:
+            f.write("Touching")
+
+        # Put it in the past, make it not writable
+        os.utime(mofile, (1000, 1000))
+        os.chmod(mofile, 0)
```
Comment (mgedmin) [inline src/zope/i18n/tests/test_compile.py]: I wonder if these tests would pass on a Windows platform...

Comment (mgedmin) [reply src/zope/i18n/tests/test_compile.py]: (I'm not sure we care, now that winbot is decommissioned.)

File: src/zope/i18n/tests/test_formats.py [hunk 19]
```diff
@@ -77,22 +90,25 @@ def getDayNames(self):
...
-        for item in self.days.items():
-            if item[1][1] == abbr:
-                return item[0]
+        raise NotImplementedError()
+
 
+class _TestCase(TestCase):
+    # Avoid deprecation warnings in Python 3 by making the preferred
+    # method name available for Python 2.
+    assertRaisesRegex = getattr(TestCase, 'assertRaisesRegex', TestCase.assertRaisesRegexp)
```
Comment (mgedmin) [inline src/zope/i18n/tests/test_formats.py]: Nice trick!

File: src/zope/i18n/tests/test_formats.py [hunk 20]
```diff
@@ -171,19 +187,23 @@ def testParseQuotesInPattern(self):
...
-        except DateTimePatternParseError as err:
-            self.assertEqual(
-                str(err), 'The quote starting at character 2 is not closed.')
+
         # Test correct length of characters in datetime fields
-        try:
-            parseDateTimePattern("HHHHH")
-        except DateTimePatternParseError as err:
-            self.assertTrue(str(err).endswith('You have: 5'))
+        parseDateTimePattern("HHHHH")
```
Comment (mgedmin) [inline src/zope/i18n/tests/test_formats.py]: Uh, I think this was supposed to fail.  Why didn't it fail?

File: src/zope/i18n/tests/test_formats.py [hunk 21]
```diff
@@ -365,11 +387,56 @@ def testParseUnusualFormats(self):
...
+        # XXX: Bug: This raises
+        # TypeError:  unsupported operand type(s) for +: 'datetime.time' and 'datetime.timedelta'
+        # at pytz/tzinfo.py:309
+        # (PyPy and CPython produce slightly different error messages;
+        # PyPy doesn't include the modulename)
+        tzinfo = pytz.timezone("US/Central")
+        with self.assertRaisesRegex(TypeError, 'timedelta'):
+            self.format.parse(
+                '21:48:01 US/Central',
+                'HH:mm:ss zzzz')
```
Comment (mgedmin) [inline src/zope/i18n/tests/test_formats.py]: I'm tempted to just fix it.  (But you're right, the fix should be in a separate PR.)

File: src/zope/i18n/compile.py (hunk 4 above)
Comment (jamadden) [reply src/zope/i18n/compile.py]: Agreed. That's just what it was doing before.

File: src/zope/i18n/compile.py (hunk 5 above)
Comment (jamadden) [reply src/zope/i18n/compile.py]: It does on Python 3, but not on Python 2.

File: src/zope/i18n/compile.py (hunk 6 above)
Comment (jamadden) [reply src/zope/i18n/compile.py]: Dependency added.

Sorry, that was a big diff for that file.

File: src/zope/i18n/compile.py (hunk 7 above)
Comment (jamadden) [reply src/zope/i18n/compile.py]: That's fair. I was just being too lazy to deal with testing the log messages.

(I guess I also don't *like* testing log messages; that feels like testing a side-effect to me.)

File: src/zope/i18n/compile.py (hunk 5 above)
Comment (mgedmin) [reply src/zope/i18n/compile.py]: Oh, it's not a locally-defined class!

File: src/zope/i18n/config.py (hunk 9 above)
Comment (jamadden) [reply src/zope/i18n/config.py]: Good point; done.

File: src/zope/i18n/format.py (hunk 10 above)
Comment (jamadden) [reply src/zope/i18n/format.py]: Yes, sorry 🙁 It could be done, but there are a lot of branches and it's based on positions in a list and it was just too much to try to come up with a *meaningful* test...

File: src/zope/i18n/locales/xmlfactory.py (hunk 14 above)
Comment (jamadden) [reply src/zope/i18n/locales/xmlfactory.py]: Hah! Yeah. copy-and-paste error. Doctest doesn't mind, though. (Which is sad, now that I think about it.)

File: src/zope/i18n/tests/test_compile.py (hunk 18 above)
Comment (jamadden) [reply src/zope/i18n/tests/test_compile.py]: As far as I can tell from [the documentation](https://docs.python.org/3/library/os.html#os.chmod) setting files to read-only should work on Windows, which is all this test cares about.

File: src/zope/i18n/tests/test_formats.py (hunk 20 above)
Comment (jamadden) [reply src/zope/i18n/tests/test_formats.py]: Good question. There's nothing in the code that produces an error message like that. The parsing function is actually quite lenient and will basically only ever produce one error (for unclosed quotes).

File: src/zope/i18n/locales/inheritance.py (hunk 13 above)
Comment (jamadden) [reply src/zope/i18n/locales/inheritance.py]: Yeah. Follow-up PR.

File: src/zope/i18n/config.py [hunk 22]
```diff
@@ -18,15 +18,26 @@ def _parse_languages(value):
...
+        >>> _parse_languages("en") == frozenset(('en',))
+        True
+        >>> _parse_languages('')
+        ''
+        >>> _parse_languages("en,es") == frozenset(('en', 'es'))
+        True
+
+    Leading, trailing and internal whitespace is ignored:
+
+        >>> _parse_languages('en, es') == frozenset(('en','es'))
```
Comment (mgedmin) [inline src/zope/i18n/config.py]: Heh.  Maybe add the space after the comma in the frozenset() literal.

File: src/zope/i18n/tests/test_formats.py (hunk 19 above)
Comment (jamadden) [reply src/zope/i18n/tests/test_formats.py]: It's amazing what you can do in one unconditional line.

File: src/zope/i18n/tests/test_formats.py (hunk 21 above)
Comment (jamadden) [reply src/zope/i18n/tests/test_formats.py]: It's not entirely clear to me how to fix it. Just don't call `tzinfo.localize` on a time?

File: src/zope/i18n/compile.py [hunk 23]
```diff
@@ -1,42 +1,36 @@
...
-    mo_mtime = 0
-    if os.path.exists(mofile):
-        # Update mo file?
-        try:
-            mo_mtime = os.stat(mofile)[ST_MTIME]
-        except (IOError, OSError):
-            return
+    if po_mtime is None:
+        logger.debug("Unable to access either %s (%s) or %s (%s)",
+                     pofile, po_mtime, mofile, mo_mtime)
```
Comment (mgedmin) [inline src/zope/i18n/compile.py]: There's no either now; it's just po_file.

File: src/zope/i18n/compile.py [hunk 24]
```diff
@@ -45,9 +39,10 @@ def compile_mo_file(domain, lc_messages_path):
             # wrap it with `closing`. Also, Msgfmt doesn't properly close a file
             # it opens for reading if you pass the path, but it does if you pass
             # the file.
-            with closing(Msgfmt(open(pofile, 'rb'), domain).getAsFile()) as mo:
-                with open(mofile, 'wb') as fd:
-                    fd.write(mo.read())
+            with open(pofile, 'rb') as pofd:
+                with closing(Msgfmt(pofd, domain).getAsFile()) as mo:
+                    with open(mofile, 'wb') as fd:
+                        fd.write(mo.read())
```
Comment (mgedmin) [inline src/zope/i18n/compile.py]: Ahh, nice readable diff.  Thank you!

File: src/zope/i18n/tests/test_formats.py (hunk 21 above)
Comment (mgedmin) [reply src/zope/i18n/tests/test_formats.py]: I was thinking `tzinfo.localize(datetime.datetime.combine(datetime.date.today(), the_parsed_time)).timetz()`.

**Comment by jamadden** [comment]
Thank you!

File: src/zope/i18n/testing.py [hunk 25]
```diff
@@ -36,8 +39,22 @@ def setUp(test=None):
     zope.component.provideAdapter(BrowserLanguages)
 
 
-class PlacelessSetup(object):
+class PlacelessSetup(CleanUp):
```
Comment (mgedmin) [inline src/zope/i18n/testing.py]: It looks like this change caused a regression: see https://github.com/zopefoundation/zope.i18n/issues/30.

I don't entirely understand why.
